Add seccomp default feature blog post

saschagrunert opened the pull request:
This adds the blog post about the new Kubernetes `SeccompDefault` alpha feature.

/hold
cc @pi-victor @pmmalinov01 @ashnehete @carlisia @chrisnegus @ritpanjw 

#### Todo

- [x] Adjust publishing date

sftim commented on content/en/blog/2021-08-15-seccomp-default.md:
```diff
@@ -0,0 +1,259 @@
+
+**Author:** Sascha Grunert, Red Hat
+
+_Editor's note: Sascha is member of [SIG Release][sigr] and is working on many
+other different container related topics. Feel free to reach him out on Twitter
+[@saschagrunert][twitter] or via the [Kubernetes slack @sascha][slack]._
+
+[sigr]: https://github.com/kubernetes/sig-release
+[twitter]: https://twitter.com/saschagrunert
+[slack]: https://kubernetes.slack.com/team/U53SUDBD4
```
We usually put contact details for SIGs rather than for blog authors.

saschagrunert replied:
Fixed and linked to the SIG Node slack.

sftim commented on content/en/blog/2021-08-15-seccomp-default.md:
```diff
@@ -0,0 +1,259 @@
+
+---
+
+Applying seccomp profiles to containerized workloads is one of the key tasks
```
What's seccomp? Users might not already know.

saschagrunert replied:
Added a note what seccomp actually is :+1:

sftim commented on content/en/blog/2021-08-15-seccomp-default.md:
```diff
@@ -0,0 +1,259 @@
+work hand in hand to create, distribute and maintain the profiles over the
+applications life-cycle.
+
+Kubernetes introduced dedicated [API fields][api] in the `SecurityContext` with
```
A reader might infer that they can `kubectl get SecurityContext` in their cluster. We should explain that `securityContext` is a field you find in Pods and in pod templates.

saschagrunert replied:
Added more explanation to the paragraph :)

sftim commented on content/en/blog/2021-08-15-seccomp-default.md:
```diff
@@ -0,0 +1,259 @@
+enhancement allowed an easier way to specify if the whole pod or a specific
+container should run as:
+
+[api]: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#seccompprofile-v1-core
```
I'd link to https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#security-context-1

saschagrunert replied:
Done :+1:

sftim commented on content/en/blog/2021-08-15-seccomp-default.md:
```diff
@@ -0,0 +1,259 @@
+feature. This means that it is disabled by default and can be enabled manually
+for every single Kubernetes node.
+
+[gate]: https://k8s.io/docs/reference/command-line-tools-reference/feature-gates
```
We prefer site-relative hyperlinks, ie `/docs/reference/command-line-tools-reference/feature-gates/`

saschagrunert replied:
Fixed :+1:

sftim commented on content/en/blog/2021-08-15-seccomp-default.md:
```diff
@@ -0,0 +1,259 @@
+
+### Trying it out
+
+If the feature is enabled on a node, then we can create a new workload like
```
nit:
```suggestion
If the feature is enabled on a node, then you can create a new workload like
```

sftim commented on content/en/blog/2021-08-15-seccomp-default.md:
````diff
@@ -0,0 +1,259 @@
+[rspec]: https://github.com/opencontainers/runtime-spec/blob/0c021c1/config-linux.md#seccomp
+
+```bash
+> CONTAINER_ID=$(sudo crictl ps -q --name=test-container)
+> sudo crictl inspect $CONTAINER_ID | jq .info.runtimeSpec.linux.seccomp
````
This isn't valid bash; omit the prompt and it would be.

saschagrunert replied:
Done :+1:

sftim commented on content/en/blog/2021-08-15-seccomp-default.md:
````diff
@@ -0,0 +1,259 @@
+> sudo crictl inspect $CONTAINER_ID | jq .info.runtimeSpec.linux.seccomp
+```
+
+```json
````
As JSON is always valid YAML, a trick: mark this as YAML and it'll format better even with the ellipsis.

saschagrunert replied:
Alright, changed to YAML. :)

sftim reviewed (commented):
This looks like a thorough explanation.

Here's some feedback on the article.

saschagrunert commented on content/en/blog/2021-08-15-seccomp-default.md:
```diff
@@ -0,0 +1,259 @@
+---
+layout: blog
+title: "Hardening Kubernetes security by enabling seccomp by default"
+date: 2021-08-15
```
@PI-Victor @pmmalinov01 @ashnehete @carlisia @chrisnegus @ritpanjw how do we decide about the date?

PI-Victor replied:
Heya, i'm not sure at this point, but we do have a sig-docs blog meeting today and i can bring this up then, do you have any preferred date?

sftim replied:
There's usually been a separate group looking at the post-release comms including blog articles, tweets, more things. I assume there still would be, and if there is then they can set dates.

I'd expect to have no more than 3 new feature articles per week, with at least a 1 day gap between them (eg Mon, Weds, Fri).

jlbutler replied:
Hi all. Yes as @sftim mentioned, there's a planned cadence of 3 feature blogs a week post-release. One or two may be scheduled on the same day as the release.

The publish dates are on the [1.22 Feature Blog tracking sheet](https://docs.google.com/spreadsheets/d/1hgFWig6YBSYORP3huwEyU0VlXGcYi6wMxjYW-swzCIY), and this one is currently scheduled for publication with the release on August 4.

Let us know if this needs a change, otherwise you can set it for then. Thanks!

jlbutler replied:
Actually, looking back it doesn't appear that we double up like this. @sftim  I'd like your input, should we update the schedule so that we have release blog on its own, then a feature blog on the three-business-day cadence from there?

saschagrunert replied:
Changed the date to `2021-08-04` for now.

reylejano replied:
> Changed the date to `2021-08-04` for now.

2021-08-04 is the scheduled release day for v1.22 and the release blog. I defer to @sftim and the 1.22 Comms team on the release date: @jlbutler , @skrishna-unix , @kunal-kushwaha , @rajula96reddy

jlbutler replied:
Hi @saschagrunert , one more quick change thanks. Based on some feedback we've reworked the blog schedule.

Please set 2021-08-20 as the publication date, based on the updated feature blog tracking sheet data. Thank you!

saschagrunert replied:
Changed it to `2021-08-20`. :+1:

sftim commented:
Looks good to review and I don't see a need for a hold (the article is future dated, to a date after the scheduled release).

/hold cancel

sftim commented:
Any more feedback or is this OK for LGTM?

PI-Victor commented:
i gave this a read way back, LGTM from my side

tengqm commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
````diff
@@ -0,0 +1,253 @@
+```
+
+Now it is possible to inspect the used seccomp profile by using
+[`ctictl`][ctictl] while investigating the containers [runtime
````
typo: crictl

saschagrunert replied:
Fixed :+1:

sftim commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,253 @@
+
+**Author:** Sascha Grunert, Red Hat
+
+_Editor's note: If you have any questions or feedback about this blog post, feel
+free to reach out via the [Kubernetes slack in #sig-node][slack]._
+
+[slack]: https://kubernetes.slack.com/messages/sig-node
+
+---
```
We usually put paragraphs like this at the end of the article.

sftim commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,253 @@
+
+---
+
+Seccomp is a security mechanism for Linux processes to filter system calls
+(syscalls) based on a set of defined rules. Applying seccomp profiles to
+containerized workloads is one of the key tasks when it comes to enhancing the
+security of the application deployment. Developers, site reliability engineers
+and infrastructure administrators have to work hand in hand to create,
+distribute and maintain the profiles over the applications life-cycle.
```
Let's mention that this article is about a change that came in the v1.22 release (that's right isn't it?)

Right now we make the reader go through quite a bit of text before they see that detail.

sftim commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,253 @@
+
+With the graduation of seccomp, nothing has changed from an overall security
+perspective, because `Unconfined` is still the default. This is totally fine if
+we look from the upgrade path and backwards compatibility perspective of
```
```suggestion
you consider this from the upgrade path and backwards compatibility perspective of
```

sftim commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
````diff
@@ -0,0 +1,253 @@
+}
+```
+
+We can see that the lower level container runtime ([CRI-O][crio-home] and
````
With the docs style guide in mind, I recommend:
```suggestion
You can see that the lower level container runtime ([CRI-O][crio-home] and
```

sftim commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,253 @@
+`SecurityContext`.
+
+The feature also works when using multiple containers within a pod, for example
+if we create a workload like this:
```
```suggestion
if you create a pod like this:
```
Why change workload to pod? Well, pods aren't fault tolerant. I don't want to give any hint that defining a pod defines a long-lived workload.

sftim commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
````diff
@@ -0,0 +1,253 @@
+      image: redis:6.2
+```
+
+Then we expect that the `test-container-nginx` runs without a seccomp profile:
````
```suggestion
then you should see that the `test-container-nginx` runs without a seccomp profile:
```

sftim commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,253 @@
+
+#### Testing the Application
+
+We have to ensure in a dedicated test environment that the application code does
```
```suggestion
If you're trying this within a dedicated test environment, you have to ensure that the application code does
```
?

sftim commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,253 @@
+
+- _Recommended_: Analyzing the code (manually or by running the application with
+  [strace][strace]) for any executed syscalls which may be blocked by the
+  default profiles. If that's the case, then we can create a custom seccomp
```
```suggestion
  default profiles. If that's the case, then you can create a custom seccomp
```

sftim commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,253 @@
+
+Thank you for reading this blog post! I hope you enjoyed to see how the usage of
+seccomp profiles has been evolved in Kubernetes over the past releases as much
+as I do. Give the `SeccompDefault` a try and feel free to reach out anytime for
```
```suggestion
as I do. On your own cluster, change the default seccomp profile to `RuntimeDefault`
(using this new feature) and see the security benefits. And, of course, feel free to reach
out any time for
```
?

sftim reviewed (commented):
Hi. Here's some feedback. It's fine to ship this without addressing the style nits I'm mentioning in this review.

saschagrunert commented:
> Hi. Here's some feedback. It's fine to ship this without addressing the style nits I'm mentioning in this review.

Thank you! I addressed all of the review comments.

sftim commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,257 @@
+---
+layout: blog
+title: "Hardening Kubernetes security by enabling seccomp by default"
```
One concern I have is how to communicate the difference between:
- you can configure your cluster so that seccomp is on by default
- Kubernetes 1.22 changed a cluster-wide default. Seccomp is enabled by default if you upgrade.

The extra challenge is to avoid confusing readers who only skim the list of articles and don't click through to read this one.

saschagrunert replied:
How about making the title more explicit: 
```suggestion
title: "Enable seccomp for all workloads with a new v1.22 alpha feature"
```

sftim commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,257 @@
+infrastructure administrators have to work hand in hand to create, distribute
+and maintain the profiles over the applications life-cycle.
+
+The [`SecurityContext`][seccontext] of Pods and Containers can be used to adjust
```
_nit_
We usually write field names how they appear in the API rather than how they appear in source code, so:
```suggestion
The [`securityContext`][seccontext] of Pods and their containers can be used to adjust
```
_or_
```suggestion
You can use the [`securityContext`][seccontext] field of Pods and their containers to adjust
```

saschagrunert replied:
Sounds good, changed as suggested :+1:

sftim commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,257 @@
+run as:
+
+[seccontext]: /docs/reference/kubernetes-api/workload-resources/pod-v1/#security-context-1
+[ga]: https://github.com/kubernetes/enhancements/blob/cccc6fa/keps/sig-node/135-seccomp/README.md#seccomp-to-ga
```
```suggestion
[ga]: https://kubernetes.io/blog/2020/08/26/kubernetes-release-1.19-accentuate-the-paw-sitive/#graduated-to-stable
```

sftim reviewed (commented):
Two more nits (sorry)

What do you think?

jlbutler commented:
Once nits are resolved, please reapply lgtm ahead of August 20 publication. Thank you!

rajula96reddy commented:
All the nits are resolved rn. Can we have apply lgtm on this?

saschagrunert commented:
@sftim PTAL

sftim commented:
/sig node
/sig security

sftim commented:
LGTM (also from https://github.com/kubernetes/website/pull/28951#issuecomment-888525651)

I'd like an accuracy signoff from either SIG Node or SIG Security.

tallclair commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,257 @@
+
+Two kubelet configuration changes have to be made to enable the feature:
+
+- set the `SeccompDefault=true` via the command line (`--feature-gates`) or the
+  [kubelet configuration][kubelet] file
+- enable the feature by adding the `--seccomp-default` command line flag or via
+  the [kubelet configuration][kubelet] file (`seccompDefault: true`)
```
nit: I would use an enumerated list here to emphasize the fact that BOTH of these need to be configured.

```suggestion
1. set the `SeccompDefault=true` via the command line (`--feature-gates`) or the
  [kubelet configuration][kubelet] file
2. enable the feature by adding the `--seccomp-default` command line flag or via
  the [kubelet configuration][kubelet] file (`seccompDefault: true`)
```

tallclair commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,257 @@
+  [strace][strace]) for any executed syscalls which may be blocked by the
+  default profiles. If that's the case, then you can create a custom seccomp
+  profile based on the default by adding the additional syscalls to the
+  `"action": "SCMP_ACT_ALLOW"` section.
```
I'd love to see a follow-up blog post that goes into more depth on how to do this. Maybe by profiling a sample application, and walking through the process of writing (and deploying) the custom profile.

saschagrunert replied:
Yeah sure, I can work on that :+1:

PushkarJ replied:
+1 on blog post

tallclair commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,257 @@
+
+- _Recommended_: Run the application against an end-to-end test suite to trigger
+  all relevant code paths with `RuntimeDefault` enabled. If a test fails, use
+  the same mitigation as mentioned above.
```
It's notoriously difficult to get this right. I would recommend a best practice for existing workloads of manually setting the profile on the workload, and then using a rolling deployment to go to production. That way, it's easier to pause and roll-back the deployment if it causes issues.

saschagrunert replied:
Alright, I added the mentioned rolling upgrade as _Recommended_ and changed the test scenario to be _Optional_.

tallclair commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,257 @@
+
+- _Recommended_: Analyzing the code (manually or by running the application with
+  [strace][strace]) for any executed syscalls which may be blocked by the
+  default profiles. If that's the case, then you can create a custom seccomp
```
I might suggest saying something like:
```suggestion
  default profiles. If that's the case, then you can override the default by explicitly setting the pod or container to run as `Unconfined`. Alternatively, you can create a custom seccomp profile (see optional step below).
```

tallclair commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,257 @@
+seccomp events. If the application works as intended then the feature can be
+enabled for further nodes within the cluster.
+
+## Conclusion
```
Are there any plans to default the flag `--secomp-default=true` in the future? I think it would be worth saying so either way in the blog post.

sftim replied:
We avoid “statements about the future” in the docs. In blog posts, it's fine to talk about these as intentions (or, if SIG Architecture agrees / would agree, as commitments).

saschagrunert replied:
Added a sentence to mention that it may be enabled per default in one of the future versions of Kubernetes.

mmdriley replied:
This new sentence seems in conflict with this (very accurate) statement earlier in the post:
> With the graduation of seccomp, nothing has changed from an overall security perspective, because `Unconfined` is still the default. This is totally fine if you consider this from the upgrade path and backwards compatibility perspective of Kubernetes releases.

And in fact this was the reasoning behind changing the way the feature worked in https://github.com/kubernetes/enhancements/pull/2773.

Raising the prospect of enabling this by default, without any new details on how cluster admins will be able to roll it out safely at scale, seems like an empty promise at best and a vague cause for alarm at worst.

saschagrunert replied:
@mmdriley yeah I get your point.

Given that the time till the blog release is only 2 days away I decided to revert the sentence and keep it as is.

mmdriley replied:
@saschagrunert thank you! I think it's a good change and the post is clearer this way.

tallclair reviewed (commented):
Thanks for writing this!

sftim commented:
[Article preview](https://deploy-preview-28951--kubernetes-io-main-staging.netlify.app/blog/2021/08/20/seccomp-default/)

PushkarJ commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,260 @@
+    - name: test-container-nginx
+      image: nginx:1.21
+      securityContext:
+        seccompProfile:
```
One thing that might be worth clarifying is how this field works for privileged containers:

Does setting `privileged: true` automatically set the `seccompProfile` to `Unconfined` or does it have to be set in addition to `privileged: true`

If it is the latter, many pod yamls that set `privileged: true` might need an update to add the `seccompProfile` field to `Unconfined`, when running on nodes with this feature enabled. 

If this is already mentioned somewhere in other docs, it would be good to link that here somewhere in the blog post.

saschagrunert replied:
Hey, the behavior does not change from the current one: Privileged containers will always run as unconfined.

saschagrunert replied:
Added a follow-up to the docs: https://github.com/kubernetes/website/pull/29489

PushkarJ replied:
Thank you for making this change

PushkarJ commented:
@saschagrunert this is a great feature spotlight blog post overall. Thank you for writing this :) 

Just added one comment requesting to clarify how this works with privileged containers. Rest seems good!

/sig security

sftim reviewed (commented):
/hold

Publication is due tomorrow and reviews are not yet in place.

If this article goes past the publication date, we should pick a new date, update the article to match that, check all is good, and then remove this hold.

saschagrunert commented:
> /hold
> 
> Publication is due tomorrow and reviews are not yet in place.
> 
> If this article goes past the publication date, we should pick a new date, update the article to match that, check all is good, and then remove this hold.

From my point of view the open discussion point has been resolved. Are we good to merge?

sftim commented:
@saschagrunert can you point to a signoff that confirms what this PR is saying is technically correct? If that were available, I'd be happy to approve it right away.

sftim commented:
@PushkarJ are you willing to add /lgtm here?

saschagrunert commented:
> @saschagrunert can you point to a signoff that confirms what this PR is saying is technically correct? If that were available, I'd be happy to approve it right away.

I only have source code, container runtimes basically do nothing on privileged containers:

https://github.com/containerd/containerd/blob/cf600abecc27200a3c0e1415cd1f6c325eb05edf/pkg/cri/server/container_create_linux.go#L424

CRI-O behaves in the same way.

sftim commented:
From the blog team's perspective, we're looking for a subject matter expert (or someone who is expert enough) to check technical accuracy. SIG Docs can't promise to have the skills for this.

PushkarJ commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,260 @@
+
+Two kubelet configuration changes have to be made to enable the feature:
+
+1. set the `SeccompDefault=true` via the command line (`--feature-gates`) or the
+   [kubelet configuration][kubelet] file
+2. enable the feature by adding the `--seccomp-default` command line flag or via
+   the [kubelet configuration][kubelet] file (`seccompDefault: true`)
```
As a reader who is not familiar with enabling this feature and feature gates, these two steps read very similar to each other. Maybe some more clarity like below would help?

1. **Enable the feature gate** `<rest of the content follows>`
2. **Turn on the feature** `<rest of the content follows>`

saschagrunert replied:
Fixed as suggested :+1:

PushkarJ commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,260 @@
+
+Please note that the feature will not influence any Kubernetes API for now.
+Therefore, it is not possible to retrieve the used seccomp profile via `kubectl`
+`get` or `describe` if the [`SeccompProfile`][api] field is unset within the
```
Is the `[api]` anchor missing?

saschagrunert replied:
Yep, re-added it.

PushkarJ commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
````diff
@@ -0,0 +1,260 @@
+true
+```
+
+Whereas the container `test-container-nginx` runs with `RuntimeDefault`:
````
Should this be `test-container-redis` ?

saschagrunert replied:
Uh, fixed.

PushkarJ commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,260 @@
+  `/var/log/audit/audit.log`) or syslog entries (defaults to `/var/log/syslog`)
+  for syscalls via `type=SECCOMP` (for audit) or `type=1326` (for syslog).
+  Compare the syscall ID with those [listed in the Linux Kernel
+  sources][syscalls] and add them to the custom profile.
```
Would an existing audit policy drop any events with `type=SECCOMP` which might lead to missing some syscalls that were `blocked`? If yes, perhaps adding that as caveat would be helpful

saschagrunert replied:
Yes, mentioned that the configuration of auditd has impact on that.

PushkarJ reviewed (changes requested):
These are mostly quick fixes and minor nits. Happy to lgtm after that

saschagrunert commented:
@PushkarJ added the changes. Thank you for the review, PTAL again.
@sftim do you have any suggestion on a new date?

PushkarJ commented:
Thank you @saschagrunert for patiently resolving all the comments. Looking forward to this blog post and more people using the feature as a result of that :) 

/lgtm

reylejano commented on content/en/blog/_posts/2021-08-20-seccomp-default.md:
```diff
@@ -0,0 +1,267 @@
+---
+layout: blog
+title: "Enable seccomp for all workloads with a new v1.22 alpha feature"
+date: 2021-08-20
```
```suggestion
date: 2021-08-25
```
Wednesday, Aug 25 is open as a publishing date

saschagrunert replied:
Changed as suggested :+1:

sftim commented:
/hold cancel
/approve

borrowing LGTM from https://github.com/kubernetes/website/pull/28951#issuecomment-903934946
/lgtm
